File: modulo08-arquitetura-de-sistemas-com-ia/modulo-04-padroes-ai-especificos/trialforge-gateway-prototype-java/src/test/java/com/trialforge/gateway/Bm25Test.java
package com.trialforge.gateway;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class Bm25Test {

    // Mesmo corpus/query dos testes puros dos dois originais: o doc que
    // compartilha "idade" e "minima" com a query deve vencer a busca lexical.
    @Test
    void docComTermosDaQueryVenceABuscaLexical() {
        List<String> corpus = List.of(
                "idade mínima de doze anos para participar do estudo",
                "consentimento do responsável legal é obrigatório",
                "retirada do participante a qualquer momento sem justificativa");
        Bm25.EstatisticasBM25 estatisticas = Bm25.construirEstatisticasBM25(corpus);
        List<String> query = Tokenizer.tokenizar("qual a idade mínima exigida");

        double melhorScore = -1;
        int vencedor = -1;
        for (int idx = 0; idx < corpus.size(); idx++) {
            double score = Bm25.scoreBM25(query, estatisticas.tokensPorDoc().get(idx), estatisticas);
            if (score > melhorScore) {
                melhorScore = score;
                vencedor = idx;
            }
        }

        assertThat(vencedor).isZero();
    }

    @Test
    void semTermoEmComumScoreEhZero() {
        List<String> corpus = List.of("gatos e cachorros", "carros e motos");
        Bm25.EstatisticasBM25 estatisticas = Bm25.construirEstatisticasBM25(corpus);
        List<String> query = Tokenizer.tokenizar("nenhumtermoemcomum");

        for (int idx = 0; idx < corpus.size(); idx++) {
            double score = Bm25.scoreBM25(query, estatisticas.tokensPorDoc().get(idx), estatisticas);
            assertThat(score).isZero();
        }
    }

    @Test
    void construirEstatisticasBM25CalculaAvgdlEDf() {
        Bm25.EstatisticasBM25 estatisticas = Bm25.construirEstatisticasBM25(List.of("a b c", "a b"));

        assertThat(estatisticas.n()).isEqualTo(2);
        assertThat(estatisticas.avgdl()).isEqualTo(2.5);
        assertThat(estatisticas.df().get("a")).isEqualTo(2);
        assertThat(estatisticas.df().get("c")).isEqualTo(1);
    }
}

File: modulo08-arquitetura-de-sistemas-com-ia/modulo-04-padroes-ai-especificos/trialforge-gateway-prototype-go/gateway/bm25_test.go
package gateway

import "testing"

// Mesmo corpus/query dos testes puros dos dois originais: o doc que
// compartilha "idade" e "minima" com a query deve vencer a busca lexical.
func TestScoreBM25(t *testing.T) {
	corpus := []string{
		"idade mínima de doze anos para participar do estudo",
		"consentimento do responsável legal é obrigatório",
		"retirada do participante a qualquer momento sem justificativa",
	}
	estatisticas := ConstruirEstatisticasBM25(corpus)
	query := Tokenizar("qual a idade mínima exigida")

	melhorScore := -1.0
	vencedor := -1
	for idx := range corpus {
		score := ScoreBM25Padrao(query, estatisticas.TokensPorDoc[idx], estatisticas)
		if score > melhorScore {
			melhorScore = score
			vencedor = idx
		}
	}

	if vencedor != 0 {
		t.Errorf("esperado doc 0 vencer a busca lexical, venceu doc %d", vencedor)
	}
}

func TestScoreBM25SemTermoComum(t *testing.T) {
	corpus := []string{"gatos e cachorros", "carros e motos"}
	estatisticas := ConstruirEstatisticasBM25(corpus)
	query := Tokenizar("nenhumtermoemcomum")

	for idx := range corpus {
		score := ScoreBM25Padrao(query, estatisticas.TokensPorDoc[idx], estatisticas)
		if score != 0 {
			t.Errorf("esperado score 0 sem termos em comum, obtido %v", score)
		}
	}
}

func TestConstruirEstatisticasBM25(t *testing.T) {
	corpus := []string{"a b c", "a b"}
	estatisticas := ConstruirEstatisticasBM25(corpus)

	if estatisticas.N != 2 {
		t.Errorf("N = %d, esperado 2", estatisticas.N)
	}
	esperadoAvgdl := 2.5
	if estatisticas.Avgdl != esperadoAvgdl {
		t.Errorf("Avgdl = %v, esperado %v", estatisticas.Avgdl, esperadoAvgdl)
	}
	if estatisticas.DF["a"] != 2 {
		t.Errorf(`DF["a"] = %d, esperado 2`, estatisticas.DF["a"])
	}
	if estatisticas.DF["c"] != 1 {
		t.Errorf(`DF["c"] = %d, esperado 1`, estatisticas.DF["c"])
	}
}
